tests/genai/evaluate/test_session_utils.py

```diff
@@ -0,0 +1,662 @@
+import os
+from unittest.mock import Mock, patch
+
+import pytest
+
+import mlflow
+from mlflow.entities import TraceData, TraceInfo, TraceLocation, TraceState
+from mlflow.entities.assessment_source import AssessmentSource, AssessmentSourceType
+from mlflow.entities.trace import Trace
+from mlflow.exceptions import MlflowException
+from mlflow.genai import scorer
+from mlflow.genai.evaluation.entities import EvalItem
+from mlflow.genai.evaluation.session_utils import (
+    classify_scorers,
+    evaluate_multi_turn_scorers,
+    get_first_trace_in_session,
+    group_traces_by_session,
+    validate_session_level_evaluation_inputs,
+)
+from mlflow.tracing.constant import TraceMetadataKey
+
+
+class _MultiTurnTestScorer:
+    """Helper class for testing multi-turn scorers."""
+
+    def __init__(self, name="test_multi_turn_scorer"):
+        self.name = name
+        self.is_session_level_scorer = True
+        self.aggregations = []
+
+    def run(self, session=None, **kwargs):
+        return True
+
+    def __call__(self, traces=None, **kwargs):
+        return 1.0
+
+
+# ==================== Tests for classify_scorers ====================
+
+
+def test_classify_scorers_all_single_turn():
+    """Test that all scorers are classified as single-turn when none are multi-turn."""
+
+    @scorer
+    def custom_scorer1(outputs):
+        return 1.0
+
+    @scorer
+    def custom_scorer2(outputs):
+        return 2.0
+
+    scorers_list = [custom_scorer1, custom_scorer2]
+    single_turn, multi_turn = classify_scorers(scorers_list)
+
+    assert len(single_turn) == 2
+    assert len(multi_turn) == 0
+    assert single_turn == scorers_list
+
+
+def test_classify_scorers_all_multi_turn():
+    """Test that all scorers are classified as multi-turn.
+
+    When all scorers have is_session_level_scorer=True.
+    """
+    multi_turn_scorer1 = _MultiTurnTestScorer(name="multi_turn_scorer1")
+    multi_turn_scorer2 = _MultiTurnTestScorer(name="multi_turn_scorer2")
+
+    scorers_list = [multi_turn_scorer1, multi_turn_scorer2]
+    single_turn, multi_turn = classify_scorers(scorers_list)
+
+    assert len(single_turn) == 0
+    assert len(multi_turn) == 2
+    assert multi_turn == scorers_list
+    # Verify they are actually multi-turn
+    assert multi_turn_scorer1.is_session_level_scorer is True
+    assert multi_turn_scorer2.is_session_level_scorer is True
+
+
+def test_classify_scorers_mixed():
+    """Test classification of mixed single-turn and multi-turn scorers."""
```

Comment:
nit: maybe don't need these one-line docstrings that are already implied by test title.